### PR TITLE
Generalized fix for sending block data to plot worker

### PR DESCRIPTION
**Pull request by cfoust**
**User-Facing Changes**
Fixes for a range of issues affecting plot preloading.

**Description**
This PR does the following:
* Breaks out the logic we use to decide when to send block data to the plot worker.
* Adds a test suite for it.

This is a complicated state machine with a few nasty edge cases. The existing logic did not handle them very well and was hard to reason about. To learn more about why this is hard, read [this comment](https://github.com/foxglove/studio/blob/d85325f895fec865bfb6c8fb8bada542224e665d/packages/studio-base/src/panels/Plot/blocks.test.ts#L45) in the test file.

I added tests for all of the edge cases I could come up with and ensured that this PR fixes the three ([#1](https://github.com/foxglove/studio/pull/6928) [#2](https://linear.app/foxglove/issue/FG-5215/strange-plot-behavior-when-used-alongside-map-panel) [#3](https://github.com/foxglove/studio/issues/6957)) recent bugs users experienced related to this functionality.

### CODE REVIEW

**Comment by cfoust** [comment]
Still investigating a couple of bugs (one seems unrelated, one probably is) but this might be worth an initial look.

**Comment by cfoust** [comment]
Scratch that. We're still clearing out the cache too aggressively; I'm investigating.

**Comment by cfoust** [comment]
I verified that all of the failing stories were still rendered correctly locally.

File: packages/studio-base/src/panels/Plot/useDatasets.ts [hunk 1]
```diff
@@ -254,57 +237,24 @@ function useData(id: string, params: PlotParams) {
...
-      void service?.addBlock(messages, Array.from(resetData));
+    const {
+      state: newState,
+      resetTopics,
+      newData,
+    } = processBlocks(blocks, blockSubscriptions, blockState);
+
+    blockState = newState;
+
+    void service?.addBlock(
```
Comment (cfoust) [inline packages/studio-base/src/panels/Plot/useDatasets.ts:248]: This is a hack so we can avoid changing the plot worker's API until after #6926 lands. After it does, we can make a `clearBlocks` function that takes a topic and clears plots that use it rather than just using the `resetTopics` parameter of `addBlocks`.

File: packages/studio-base/src/panels/Plot/blocks.test.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+// I came up with a concise way of describing this problem to make the state
+// machine's behavior easier to understand. Message blocks can be represented
+// as a 1D array of cells, each of which can be in one of four states:
+// s = same
+// c = changed
+// n = new
+// e = empty
+//
+// Blocks are accumulated and sent using a cursor system, here represented by
+// the pipe character ("|"). This indicates that all of the data prior to the
```
Comment (snosenzo) [inline packages/studio-base/src/panels/Plot/blocks.test.ts:58]: I see this, but I also see examples below with `n`'s and `s`'s after the cursor. Why wouldn't the cursor consume these? Shouldn't the cursor always push to the end or the latest empty block?

File: packages/studio-base/src/panels/Plot/blocks.test.ts [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+// sssssss|eee
+//
+// In instances where the data should be cleared, we append an asterisk ("*")
+// to the end of the state string.
+//
+// Some examples:
+// ssss|nnneee -> sssssss|eee (data accumulating normally)
+// s|nene -> sses|e (skipping empty data)
+// cs|nnneee -> s|snnneee* (a block was changed before the cursor)
+// sss|ccsss -> sssss|sss (a block was changed after the cursor)
```
Comment (snosenzo) [inline packages/studio-base/src/panels/Plot/blocks.test.ts:85]: When would a case like this ever happen? For the user script case, wouldn't all of the blocks be changed or none of them?

Comment (cfoust) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:85]: yup, it happens regularly, both when a user script changes and when slicing does. the blocks are not all changed at once.

File: packages/studio-base/src/panels/Plot/blocks.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+import { MessageBlock } from "@foxglove/studio-base/PanelAPI/useBlocksSubscriptions";
+import { SubscribePayload } from "@foxglove/studio-base/players/types";
+
+import { Messages } from "./internalTypes";
+
+// We need to keep track of the block data we've already sent to the worker and
+// detect when it has changed, which can happen when the user changes a user
+// script or they trigger a subscription to different fields.
+// mapping from topic -> the first message on that topic in the block
+type FirstMessages = Record<string, unknown>;
```
Comment (snosenzo) [inline packages/studio-base/src/panels/Plot/blocks.ts:16]: `FirstMessagesByTopic` would let you remove these `topic->  ` clarifying comments, same for `Cursors`

File: packages/studio-base/src/panels/Plot/blocks.test.ts (hunk 2 above)
Comment (cfoust) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:58]: Any `n` after the cursor will be consumed regardless. Since we only consume up to and including the last changed block in the event that changes occur after the cursor, it is not desirable for us to consume all `s`.

The reason for this is for the case where a block is updated before the cursor, but blocks after the changed block are untouched. This occurs when a user script is nondeterministic (i.e. what occurred in issue [#1](https://github.com/foxglove/studio/pull/6928)). We can't know whether blocks will continue to change after the changed blocks, so we have to wait until the next iteration to be sure.

To illustrate:
```
sscs| -> sss|s*
sss|s -> ssss|
```

File: packages/studio-base/src/panels/Plot/blocks.test.ts (hunk 3 above)
Comment (snosenzo) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:85]: What do we gain by using cursors then if changes can happen anywhere in the array? Why not just use array status diffs for blocks if they need to be iterated through anyways?

File: packages/studio-base/src/panels/Plot/blocks.test.ts (hunk 2 above)
Comment (snosenzo) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:58]: seems strange to me that the cursor position doesn't always represent that data before the cursor has already be sent. Why can't/shouldn't it resolve that over one iteration?

Comment (cfoust) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:58]: The cursor position always represents the data that has already been sent. That's not violated anywhere. In the case I described above, we cannot know whether data will change following the cursor, so we don't consume it until the next iteration.

File: packages/studio-base/src/panels/Plot/blocks.test.ts (hunk 3 above)
Comment (cfoust) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:85]: We need to keep track of the data we've already sent to handle (a) changes that occur before the cursor and (b) data that is still streaming in.

File: packages/studio-base/src/panels/Plot/blocks.ts (hunk 4 above)
Comment (cfoust) [reply packages/studio-base/src/panels/Plot/blocks.ts:16]: Hmm, perhaps--personally I prefer specifying the mapping separately for associative arrays to keep identifiers short.

Comment (snosenzo) [reply packages/studio-base/src/panels/Plot/blocks.ts:16]: The short identifiers are sacrificing readability and incur more mental overhead when reviewing imo

Comment (cfoust) [reply packages/studio-base/src/panels/Plot/blocks.ts:16]: I see why this is important to you, but everyone brings a different perspective on naming. Unless a name is blatantly incorrect or misleading, I think we should concentrate our efforts on examining the correctness of the methodology of PRs, not their idiosyncrasies

File: packages/studio-base/src/panels/Plot/blocks.test.ts (hunk 2 above)
Comment (snosenzo) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:58]: What triggers these iterations? Is it safe to assume that there will always be another iteration that will consume the whole array?

Comment (cfoust) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:58]: Any reference change to `blocks` returned [here](https://github.com/foxglove/studio/pull/6956/files#diff-7372e3a2054c7c5fd4edb8e904c174c725b3ed43c034938df26f5a442918c881R218) will trigger an iteration. In practice the player (or more specifically, the hook) will always bump the reference if data is being added or changed.

File: packages/studio-base/src/panels/Plot/blocks.test.ts (hunk 3 above)
Comment (snosenzo) [reply packages/studio-base/src/panels/Plot/blocks.test.ts:85]: I think something that wasn't really clear to me before reading this was that we can only really add or reset blocks as an outcome to this process. This context is important to know for why we're using a cursor. Those requirements aren't really said anywhere.

**Review by snosenzo — APPROVED** [review]
would be good if @jtbandes to give a look over too. tested and works.

I still think that the cursor behavior that assumes there will be another block update trigger after not consuming the entire array of blocks in this example (see below) is a bit fragile and over-complex, but didn't run into any related issues testing against user scripts. 
```
sscs| -> sss|s*
sss|s -> ssss|
```
